vcu118/init.c

```diff
@@ -3,12 +3,30 @@
 
 extern int main(void);
 
+#define write_csr(reg, val) \
+  asm volatile ("csrw " #reg ", %0" :: "r"(val))
+
+#define read_csr(reg) ({ unsigned long __tmp;               \
+            asm volatile ("csrr %0, " #reg : "=r"(__tmp));  \
+            __tmp; })
+
+void bad_trap(){
+  unsigned long mepc, mtval, mcause;
+  asm volatile("csrr %0, mcause" : "=r"(mcause));
```

Comment:
Why not use `read_csr` here (and the below lines)?

Reply:
No reason, I could ... less code being generated this way though.